src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -362,106 +365,100 @@ private[concurrent] object Promise {
-    // Invariant: _arg is `ExecutionContext`, and non-null. `this` ne Noop.
+    // Invariant: _arg is `null`, _ec is non-null. `this` ne Noop.
     // requireNonNull(resolved) will hold as guarded by `resolve`
     final def submitWithValue(resolved: Try[F]): this.type = {
-      val e = _arg.asInstanceOf[ExecutionContext]
       _arg = resolved
-      try e.execute(this) /* Safe publication of _arg and _fun */
+      val e = _ec
+      try e.execute(this) /* Safe publication of _arg, _fun, _ec */
       catch {
-        case t: Throwable => handleFailure(t, e)
+        case t: Throwable =>
+          _fun = null // allow to GC
+          _arg = null // see above
+          _ec  = null // see above again
+          handleFailure(t, e)
       }
+
       this
     }
 
     private[this] final def handleFailure(t: Throwable, e: ExecutionContext): Unit = {
-      _fun = null // allow to GC
-      _arg = null // see above
       val wasInterrupted = t.isInstanceOf[InterruptedException]
-      if (NonFatal(t) || wasInterrupted) {
+      if (wasInterrupted || NonFatal(t)) {
         val completed = tryComplete0(get(), resolve(Failure(t)))
         if (completed && wasInterrupted) Thread.currentThread.interrupt()
-        if (!completed && (e ne null)) e.reportFailure(t)
+
+        // Report or rethrow failures which are unlikely to otherwise be noticed
+        if (_xform == Xform_foreach || _xform == Xform_onComplete || !completed)
+          e.reportFailure(t)
       } else throw t
     }
 
     // Gets invoked by the ExecutionContext, when we have a value to transform.
-    // Invariant: if (_arg.isInstanceOf[Try[F]] && (_fun ne null))
-    override final def run(): Unit =
+    override final def run(): Unit = {
+      val v   = _arg
+      val fun = _fun
+      val ec  = _ec
+      _fun = null // allow to GC
+      _arg = null // see above
+      _ec  = null // see above
       try {
-        val v = _arg.asInstanceOf[Try[F]]
-        (_xform.toInt: @switch) match {
-          case Xform_noop          => doAbort(v)
-          case Xform_map           => doMap(v)
-          case Xform_flatMap       => doFlatMap(v)
-          case Xform_transform     => doTransform(v)
-          case Xform_transformWith => doTransformWith(v)
-          case Xform_foreach       => v.foreach(_fun)
-          case Xform_onComplete    => _fun(v)
-          case Xform_recover       => doRecover(v)
-          case Xform_recoverWith   => doRecoverWith(v)
-          case Xform_filter        => doFilter(v)
-          case Xform_collect       => doCollect(v)
-          case _                   => doAbort(v)
-        }
-        _fun = null // allow to GC
-        _arg = null // see above
+        val resolvedResult: Try[_] =
+          (_xform: @switch) match {
+            case Xform_noop          =>
+              null
+            case Xform_map           =>
+              if (v.isInstanceOf[Success[F]]) Success(fun(v.get)) else v // Faster than `resolve(v map fun)`
```

Comment:
@NthPortal Yes, it's handled here: https://github.com/scala/scala/pull/7663/files#diff-e0c06059ba197fea5927c4e4b0f3d837R398